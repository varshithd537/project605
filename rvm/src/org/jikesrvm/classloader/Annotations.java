/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm.classloader;

public class Annotations {

  private static final Annotations NO_ANNOTATIONS = new Annotations(new RVMAnnotation[0]);

  protected final RVMAnnotation[] annotations;

  protected byte[] rawAnnotations;

  /**
   * @param annotations array of runtime visible annotations
   */
  public Annotations(RVMAnnotation[] annotations) {
    this.annotations = annotations;
  }

  public RVMAnnotation[] getAnnotations() {
    return annotations;
  }

  static Annotations noAnnotations() {
    return NO_ANNOTATIONS;
  }

  public void setRawAnnotations(byte[] rawAnnotations) {
    this.rawAnnotations = rawAnnotations;
  }

  public byte[] getRawAnnotations() {
    return rawAnnotations;
  }
}
