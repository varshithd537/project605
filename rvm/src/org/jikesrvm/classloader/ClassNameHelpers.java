/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm.classloader;

public final class ClassNameHelpers {

  /**
   * Converts a normal class name to an internal binary class name.
   * @param className a class name
   * @return the internal binary form of the class name
   *
   * @see <a href="https://docs.oracle.com/javase/specs/index.html">
   * Java Virtual Machine Specification, Section "Binary Class and Interface Names"</a>
   */
  public static String convertClassnameToInternalName(String className) {
    if (className == null) {
      return null;
    }
    return className.replace('.', '/');
  }

}
