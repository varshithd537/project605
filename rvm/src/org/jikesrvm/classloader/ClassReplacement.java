/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm.classloader;

import org.jikesrvm.VM;
import org.jikesrvm.classlibrary.AbstractReplacementClasses;
import org.jikesrvm.util.HashMapRVM;
import org.jikesrvm.util.HashSetRVM;
import org.jikesrvm.util.ImmutableEntryHashSetRVM;

/**
 * Provides methods to implement replacement of methods and fields of classes.
 */
public class ClassReplacement {

  private static final String REPLACEMENT_CLASS_NAME_PREFIX = "org.jikesrvm.classlibrary.openjdk.replacements.";

  private AbstractReplacementClasses replacementClasses;
  private HashSetRVM<String> remainingReplacementClassNames;
  private HashMapRVM<String, String> mapOfTargetClassToSourceClass;

  private final ImmutableEntryHashSetRVM<String> loadedReplacementClasses = new ImmutableEntryHashSetRVM<String>();
  private final ImmutableEntryHashSetRVM<String> classesCheckedForReplacements = new ImmutableEntryHashSetRVM<String>();

  private final BootstrapClassLoader bootstrapCL;

  public ClassReplacement(BootstrapClassLoader bootstrapCL) {
    this.bootstrapCL = bootstrapCL;
    if (VM.BuildForOpenJDK) {
      try {
        this.replacementClasses =       (AbstractReplacementClasses)Class.forName("org.jikesrvm.classlibrary.ReplacementClasses").newInstance();
      } catch (ExceptionInInitializerError e) {
        throw new Error("Throwable during construction of BootstrapClassLoader", e);
      } catch (SecurityException e) {
        throw new Error("Throwable during construction of BootstrapClassLoader", e);
      } catch (IllegalAccessException e) {
        throw new Error("Throwable during construction of BootstrapClassLoader", e);
      } catch (InstantiationException e) {
        throw new Error("Throwable during construction of BootstrapClassLoader", e);
      } catch (ClassNotFoundException e) {
        throw new Error("Throwable during construction of BootstrapClassLoader", e);
      }
      remainingReplacementClassNames = replacementClasses.getNamesOfClassesWithReplacements();
      mapOfTargetClassToSourceClass = replacementClasses.getMapOfTargetClassToSourceClass();
    } else {
      remainingReplacementClassNames = new HashSetRVM<String>(1);
      mapOfTargetClassToSourceClass = new HashMapRVM<String, String>(1);
    }
  }

  public void attemptToLoadReplacementClassIfNeededForVmClass(String className)
      throws NoClassDefFoundError, ClassNotFoundException {
    if (remainingReplacementClassNames.contains(className)) {
      if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: " + className + " has a replacement!");
    } else {
      if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: " + className + " has NO replacement!");
      return;
    }

    Atom classNameAtom = null;
    String replacementClassName = null;
    try {
      // TODO OPENJDK/ICEDTEA Adopt better nomenclature for replacement classes
      classNameAtom = Atom.findOrCreateAsciiAtom(className);
      replacementClassName = toReplacementClassNameAtom(className);
      if (VM.VerifyAssertions) VM._assert(replacementClassName != null);
      // TODO OPENJDK/ICEDTEA needs a method to convert class names to descriptors and vice versa. possibly in atom if it doesn't exist yet
      Atom replacementClassDescriptor = Atom.findOrCreateAsciiAtom("L" + replacementClassName.toString().replace('.', '/') + ";");

      boolean isReplacementClass = className.startsWith(REPLACEMENT_CLASS_NAME_PREFIX);

      if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: " + className + " is a replacement class itself and thus won't be checked for a replacement class: " + isReplacementClass);
      if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: " + className + " is a bootstrap class descriptor? " + replacementClassDescriptor.isBootstrapClassDescriptor());
      if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: " + className + " has been checked for replacement? " + classesCheckedForReplacements.contains(className));
      if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: " + className + " would have replacement class name: " + replacementClassName);
      if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: " + className + " would have replacement class descriptor: " + replacementClassDescriptor);

      boolean alreadyTriedToLoadReplacementClass = classesCheckedForReplacements.contains(className);
      // Check if there has been an attempt to load the replacement JDK class of this name
      if (!isReplacementClass && !alreadyTriedToLoadReplacementClass) {
        classesCheckedForReplacements.add(className);
        if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: Checking for replacement class for " + className + " named " + replacementClassName);

        if (VM.VerifyAssertions) VM._assert(!VM.runningVM);
        // Load, resolve and instantiate the replacement class if it exists.
        if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: About to call loadVMClass for replacement class for " + className + " named " + replacementClassName);
        RVMType loadVMClass = bootstrapCL.loadVMClass(replacementClassName.toString());
        if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: Finished with call loadVMClass for replacement class for " + className + " named " + replacementClassName);
        loadVMClass.resolve();
        loadedReplacementClasses.add(replacementClassName);
        remainingReplacementClassNames.remove(className);
        if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: Replacement class " + replacementClassName + " loaded for " + className);
      } else {
        if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: NOT LOADING replacement class for " + className + " named " + replacementClassName);
        if (VM.TraceClassLoading) VM.sysWriteln("alreadyTriedToLoadReplacementClass: " + alreadyTriedToLoadReplacementClass);
        if (VM.TraceClassLoading) VM.sysWriteln("is replacement class: " + isReplacementClass);
      }
    } catch (NoSuchMethodError e) {
        if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: Failed to load replacement class " + replacementClassName + " for " + classNameAtom);
        if (VM.TraceClassLoading) VM.sysWriteln(e.toString());
        if (VM.TraceClassLoading) e.printStackTrace();
        throw new ClassNotFoundException("Failed to load replacement class " + replacementClassName, e);
    } catch (NoClassDefFoundError e) {
      if (VM.TraceClassLoading) VM.sysWriteln("ClassReplacement_VMClass: Failed to load replacement class " + replacementClassName + " for " + classNameAtom);
      if (VM.TraceClassLoading) VM.sysWriteln(e.toString());
      if (VM.TraceClassLoading) e.printStackTrace();
    }
  }

  public String toReplacementClassNameAtom(String className) {
    return mapOfTargetClassToSourceClass.get(className);
  }

}
