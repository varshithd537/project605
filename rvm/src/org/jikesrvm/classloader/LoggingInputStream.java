/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm.classloader;

import java.io.ByteArrayOutputStream;
import java.io.FilterInputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 * Provides an implementation of a stream that can log bytes that are read.
 * <p>
 * The read bytes are logged to an internal buffer.
 */
public class LoggingInputStream extends FilterInputStream {

  private ByteArrayOutputStream out;
  private boolean loggingEnabled;

  protected LoggingInputStream(InputStream in) {
    super(in);
    this.out = new ByteArrayOutputStream();
  }

  @Override
  public int read() throws IOException {
    int read = super.read();
    if (loggingEnabled && read != -1) {
      out.write((byte) read);
    }
    return read;
  }

  @Override
  public int read(byte[] b, int off, int len) throws IOException {
    int readBytes = super.read(b, off, len);
    if (loggingEnabled && readBytes > 0) {
      out.write(b, off, readBytes);
    }
    return readBytes;
  }

  public void startLogging() {
    this.loggingEnabled = true;
  }

  public void stopLogging() {
    this.loggingEnabled = false;
  }

  public byte[] getLoggedBytes() {
    return out.toByteArray();
  }

  public void clearLoggedBytes() {
    out = new ByteArrayOutputStream();
  }

}
