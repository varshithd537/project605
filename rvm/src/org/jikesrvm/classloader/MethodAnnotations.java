/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.jikesrvm.classloader;

public class MethodAnnotations extends Annotations {

  private static final MethodAnnotations NO_METHOD_ANNOTATIONS = new MethodAnnotations(null, null, null);

  private final RVMAnnotation[][] parameterAnnotations;
  private final Object annotationDefaults;

  private byte[] rawParameterAnnotations;
  private byte[] rawAnnotationDefault;

  /**
   * @param annotations array of runtime visible annotations
   * @param parameterAnnotations array of runtime visible parameter annotations
   * @param annotationDefaults value for this annotation that appears
   */
  public MethodAnnotations(RVMAnnotation[] annotations,
      RVMAnnotation[][] parameterAnnotations, Object annotationDefaults) {
        super(annotations);
        this.parameterAnnotations = parameterAnnotations;
        this.annotationDefaults = annotationDefaults;
  }

  public RVMAnnotation[][] getParameterAnnotations() {
    return parameterAnnotations;
  }

  public Object getAnnotationDefaults() {
    return annotationDefaults;
  }

  public void setRawParameterAnnotations(byte[] rawParameterAnnotations) {
    this.rawParameterAnnotations = rawParameterAnnotations;

  }

  public void setRawAnnotationDefault(byte[] rawAnnotationDefault) {
    this.rawAnnotationDefault = rawAnnotationDefault;
  }

  public byte[] getRawParameterAnnotations() {
    return rawParameterAnnotations;
  }

  public byte[] getRawAnnotationDefault() {
    return rawAnnotationDefault;
  }

  static MethodAnnotations noMethodAnnotations() {
    return NO_METHOD_ANNOTATIONS;
  }

}
