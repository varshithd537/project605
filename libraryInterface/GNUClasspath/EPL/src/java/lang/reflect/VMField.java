/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang.reflect;

import java.lang.annotation.Annotation;

import org.jikesrvm.VM;
import org.jikesrvm.classlibrary.JavaLangReflectSupport;
import org.jikesrvm.classloader.Atom;
import org.jikesrvm.classloader.RVMClass;
import org.jikesrvm.classloader.RVMField;

/**
 * Implementation of java.lang.reflect.Field for JikesRVM.<p>
 *
 * By convention, order methods in the same order
 * as they appear in the method summary list of Sun's 1.4 Javadoc API.
 */
public final class VMField {

  final RVMField field;

  // For use by JikesRVMSupport
  VMField(RVMField f) {
    field = f;
  }

  @Override
  public boolean equals(Object object) {
    if (object instanceof Field) {
      return field == ((Field)object).f.field;
    } else {
      return false;
    }
  }

  Class<?> getDeclaringClass() {
    return field.getDeclaringClass().getClassForType();
  }

  Object get(Field f, Object object) throws IllegalAccessException, IllegalArgumentException {
    if (VM.ExtremeAssertions) {
      boolean invalidGet = JikesRVMSupport.getFieldOf(f).getType().isMagicType();
      if (invalidGet) {
        String msg = "Attempted to access reflection field for " + f +
            " which has a magic type and thus doesn't support reflection!";
        VM._assert(VM.NOT_REACHED, msg);
      }
    }

    return JavaLangReflectSupport.get(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  boolean getBoolean(Field f,Object object) throws IllegalAccessException, IllegalArgumentException {
    return JavaLangReflectSupport.getBoolean(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  byte getByte(Field f, Object object) throws IllegalAccessException, IllegalArgumentException {
    return JavaLangReflectSupport.getByte(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  char getChar(Field f, Object object) throws IllegalAccessException, IllegalArgumentException {
    return JavaLangReflectSupport.getChar(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  double getDouble(Field f, Object object) throws IllegalAccessException, IllegalArgumentException {
    return JavaLangReflectSupport.getDouble(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  float getFloat(Field f, Object object) throws IllegalAccessException, IllegalArgumentException {
    return JavaLangReflectSupport.getFloat(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  int getInt(Field f, Object object) throws IllegalAccessException, IllegalArgumentException {
    return JavaLangReflectSupport.getInt(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  long getLong(Field f, Object object) throws IllegalAccessException, IllegalArgumentException {
    return JavaLangReflectSupport.getLong(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  public short getShort(Field f, Object object) throws IllegalAccessException, IllegalArgumentException {
    return JavaLangReflectSupport.getShort(object, field, f, RVMClass.getClassFromStackFrame(2));
  }

  int getModifiersInternal() {
    return field.getModifiers();
  }

  String getName() {
    return field.getName().toString();
  }

  Class<?> getType() {
    return field.getType().resolve().getClassForType();
  }

  void set(Field f, Object object, Object value)
    throws IllegalAccessException, IllegalArgumentException     {
    if (VM.ExtremeAssertions) {
      boolean invalidSet = JikesRVMSupport.getFieldOf(f).getType().isMagicType();
      if (invalidSet) {
        String msg = "Attempted to access reflection field for " + f +
            " which has a magic type and thus doesn't support reflection!";
        VM._assert(VM.NOT_REACHED, msg);
      }
    }

    JavaLangReflectSupport.set(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

  void setBoolean(Field f, Object object, boolean value)
    throws IllegalAccessException, IllegalArgumentException    {
    JavaLangReflectSupport.setBoolean(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

   void setByte(Field f, Object object, byte value)
     throws IllegalAccessException, IllegalArgumentException    {
     JavaLangReflectSupport.setByte(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

  void setChar(Field f, Object object, char value)
    throws IllegalAccessException, IllegalArgumentException    {
    JavaLangReflectSupport.setChar(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

  void setDouble(Field f, Object object, double value)
    throws IllegalAccessException, IllegalArgumentException    {
    JavaLangReflectSupport.setDouble(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

  void setFloat(Field f, Object object, float value)
    throws IllegalAccessException, IllegalArgumentException    {
    JavaLangReflectSupport.setFloat(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

  void setInt(Field f, Object object, int value)
    throws IllegalAccessException, IllegalArgumentException    {
    JavaLangReflectSupport.setInt(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

  void setLong(Field f, Object object, long value)
    throws IllegalAccessException, IllegalArgumentException    {
    JavaLangReflectSupport.setLong(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

  void setShort(Field f, Object object, short value)
    throws IllegalAccessException, IllegalArgumentException   {
    JavaLangReflectSupport.setShort(object, value, field, f, RVMClass.getClassFromStackFrame(2));
  }

  // AnnotatedElement interface

  Annotation[] getDeclaredAnnotations() {
    return field.getDeclaredAnnotations();
  }

  <T extends Annotation> T getAnnotation(Class<T> annotationClass) {
    return field.getAnnotation(annotationClass);
  }

  String getSignature() {
    Atom signature = field.getSignature();
    return (signature == null) ? null : signature.toString();
  }
}
