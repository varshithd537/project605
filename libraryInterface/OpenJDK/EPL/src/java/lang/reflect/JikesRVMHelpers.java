/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang.reflect;

import java.io.UTFDataFormatException;

import org.jikesrvm.VM;
import org.jikesrvm.classloader.Atom;
import org.jikesrvm.classloader.RVMClass;
import org.jikesrvm.classloader.RVMField;
import org.jikesrvm.classloader.RVMMethod;
import org.jikesrvm.classloader.RVMType;
import org.jikesrvm.classloader.TypeReference;

public class JikesRVMHelpers {

  public static RVMField findFieldByName(RVMClass typeForClass, String fieldNameStr) {
    Atom fieldName = Atom.findUnicodeAtom(fieldNameStr);
    if (VM.VerifyAssertions) VM._assert(fieldName != null);
    return typeForClass.findDeclaredField(fieldName);
  }

  public static String convertAtomToInternedStringOrNull(Atom a) {
    if (a == null) {
      return null;
    }
    return atomToInternedStringOrError(a);
  }

  public static String atomToInternedStringOrError(Atom a) {
    try {
      return a.toUnicodeString();
    } catch (UTFDataFormatException e) {
      throw new Error(e);
    }
  }

  public static Class[] convertMethodParametersTypesToClasses(RVMMethod m) {
    TypeReference[] parameterTypes = m.getParameterTypes();
    Class[] classes = new Class[parameterTypes.length];
    for (int i = 0; i < parameterTypes.length; i++) {
      TypeReference typeReference = parameterTypes[i];
      RVMType type = typeReference.peekType();
      if (type == null) {
        type = typeReference.resolve();
      }
      classes[i] = type.getClassForType();
    }
    return classes;
  }

}
