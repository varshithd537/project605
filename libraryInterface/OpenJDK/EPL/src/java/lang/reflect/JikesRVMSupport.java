/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang.reflect;
import org.jikesrvm.VM;
import org.jikesrvm.classlibrary.ClassLibraryHelpers;
import org.jikesrvm.classlibrary.JavaLangReflectSupport;
import org.jikesrvm.classloader.*;
import org.jikesrvm.runtime.Reflection;
import org.jikesrvm.runtime.ReflectionBase;

import static java.lang.reflect.JikesRVMHelpers.*;

import java.util.LinkedList;
import java.util.List;

/**
 * Library support interface of Jikes RVM
 */
public class JikesRVMSupport {

  private static final ReflectAccess ra = new ReflectAccess();

  /**
   * The implementations of {@link Constructor}, {@link Method} and {@link Field}
   * have a field with name {@code slot} that has an implementation defined
   * meaning. It's not useful for us because we currently don't use shared tables
   * for fields and methods. Therefore, save some defined value into that we'll possibly
   * recognize if it appears somewhere.
   */
  public static final int SLOT_CONTENTS = 0xFADEDEAD;

  public static Field createField(RVMField f) {
    RVMClass declaringRvmClass = f.getDeclaringClass();
    Class<?> declaringClass = declaringRvmClass.getClassForType();
    String name = atomToInternedStringOrError(f.getName());
    Class<?> type = f.getType().resolve().getClassForType();
    int modifiers = f.getModifiers();
    int slot = SLOT_CONTENTS;
    String signature = convertAtomToInternedStringOrNull(f.getSignature());
    byte[] annotations = f.getAnnotationsData().getRawAnnotations();
    Field newField = ra.newField(declaringClass, name, type, modifiers, slot, signature, annotations);
    ClassLibraryHelpers.javaLangReflectField_rvmFieldField.setObjectValueUnchecked(newField, f);
    return newField;
  }

  public static Method createMethod(RVMMethod m) {
    Class<?> declaringClass = m.getDeclaringClass().getClassForType();
    String name = atomToInternedStringOrError(m.getName());
    Class[] parameterTypes = convertMethodParametersTypesToClasses(m);
    Class<?> returnType = m.getReturnType().resolve().getClassForType();
    Class[] checkedExceptions = determineCheckedExceptionTypes(m.getExceptionTypes());
    int modifiers = m.getModifiers();
    int slot = SLOT_CONTENTS;
    String signature = convertAtomToInternedStringOrNull(m.getSignature());
    MethodAnnotations annotationsData = (MethodAnnotations) m.getAnnotationsData();
    byte[] annotations = annotationsData.getRawAnnotations();
    byte[] parameterAnnotations = annotationsData.getRawParameterAnnotations();
    byte[] annotationDefault = annotationsData.getRawAnnotationDefault();
    Method newMethod = ra.newMethod(declaringClass, name, parameterTypes, returnType, checkedExceptions, modifiers, slot, signature, annotations, parameterAnnotations, annotationDefault);
    ClassLibraryHelpers.javaLangReflectMethod_rvmMethodField.setObjectValueUnchecked(newMethod, m);
    ReflectionBase invoker = m.getInvoker();
    if (Reflection.cacheInvokerInJavaLangReflect) {
      ClassLibraryHelpers.javaLangReflectMethod_invokerField.setObjectValueUnchecked(newMethod, invoker);
    } else {
      ClassLibraryHelpers.javaLangReflectMethod_invokerField.setObjectValueUnchecked(newMethod, null);
    }
    return newMethod;
  }



  @SuppressWarnings("unchecked") // Can't type-check this without <T> type<T>, which breaks javac
  public static <T> Constructor<T> createConstructor(RVMMethod m) {
    Class<?> declaringClass = m.getDeclaringClass().getClassForType();
    Class[] parameterTypes = convertMethodParametersTypesToClasses(m);
    Class[] checkedExceptions = determineCheckedExceptionTypes(m.getExceptionTypes());
    int modifiers = m.getModifiers();
    // slot is implementation defined
    int slot = SLOT_CONTENTS;
    String signature = convertAtomToInternedStringOrNull(m.getSignature());
    MethodAnnotations annotationsData = (MethodAnnotations) m.getAnnotationsData();
    byte[] annotations = annotationsData.getRawAnnotations();
    byte[] parameterAnnotations = annotationsData.getRawParameterAnnotations();
    Constructor<T> newConstructor = (Constructor<T>) ra.newConstructor(declaringClass, parameterTypes, checkedExceptions, modifiers, slot, signature, annotations, parameterAnnotations);
    ClassLibraryHelpers.javaLangReflectConstructor_rvmMethodField.setObjectValueUnchecked(newConstructor, m);
    return newConstructor;
  }

  public static Object createVMConstructor(RVMMethod m) {
    VM.sysWriteln("CreateVMConstructor is called");
    throw new Error("OpenJDK doesn't provide the VMConstructor API and shouldn't need it");
  }

  public static RVMField getFieldOf(Field f) {
    return (RVMField) ClassLibraryHelpers.javaLangReflectField_rvmFieldField.getObjectUnchecked(f);
  }

  public static RVMMethod getMethodOf(Method m) {
    return (RVMMethod) ClassLibraryHelpers.javaLangReflectMethod_rvmMethodField.getObjectUnchecked(m);
  }

  public static RVMMethod getMethodOf(Constructor cons) {
    return (RVMMethod) ClassLibraryHelpers.javaLangReflectConstructor_rvmMethodField.getObjectUnchecked(cons);
  }

  private static Class[] determineCheckedExceptionTypes(
      TypeReference[] exceptionTypeReferences) {
    // TODO OPENJDK/ICEDTEA partly sharable with GNU Classpath
    if (exceptionTypeReferences == null) {
      return new Class[0];
    } else {
      // FIXME OPENJDK/ICEDTEA move this into a shared helper or core VM and add test cases
      Class<?>[] classes = JavaLangReflectSupport.typesToClasses(exceptionTypeReferences);
      return determineCheckedExceptionTypes(classes);
    }
  }

  private static Class[] determineCheckedExceptionTypes(Class<?>[] classes) {
    List<Class<?>> checkedExceptionClasses = new LinkedList<Class<?>>();
    Class<?> ancestor = null;
    for (Class<?> c : classes) {
      ancestor = c;
      while (ancestor != null) {
        String name = ancestor.getName();
        if ("java.lang.RuntimeException".equals(name) || "java.lang.Error".equals(name)) {
          break;
        } else if ("java.lang.Exception".equals(name)) {
          checkedExceptionClasses.add(c);
          break;
        } else {
          ancestor = ancestor.getSuperclass();
        }
      }
    }
    return checkedExceptionClasses.toArray(new Class[0]);
  }


}
