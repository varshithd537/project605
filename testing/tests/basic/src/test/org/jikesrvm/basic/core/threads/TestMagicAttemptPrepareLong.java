/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.threads;

import org.jikesrvm.classloader.RVMField;
import org.jikesrvm.runtime.Magic;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

import org.vmmagic.unboxed.Offset;

class TestMagicAttemptPrepareLong extends XThread {

  public static void main(String[] args) {
    LongField lf = new LongField();
    for (int i = 0; i < 5; i++) {
      TestMagicAttemptPrepareLong tmapl = new TestMagicAttemptPrepareLong(i,lf);
      tmapl.start();
    }
    XThread.say("bye");
    XThread.outputMessages();
  }

  static volatile int vi = 0;

  int n;
  long l;
  LongField lf;

  TestMagicAttemptPrepareLong(int i, LongField lf) {
    super("MAPL" + i);
    n = i;
    l = (((long) n) << 32) + n;
    this.lf = lf;
  }

  void performTask() {
    int errors = 0;

    RVMField internalLongField = null;
    try {
      Field longField = LongField.class.getField("longAccessedViaMagic");
      Class<?> jikesRVMSupportClass = Class.forName("java.lang.reflect.JikesRVMSupport");
      Method m = jikesRVMSupportClass.getMethod("getFieldOf", Class.forName("java.lang.reflect.Field"));
      internalLongField = (RVMField) m.invoke(null, longField);
    } catch (Exception ex) {
      ex.printStackTrace();
      return;
    }

    Offset fieldOffset = internalLongField.getOffset();

    for (int i = 0; i < 10000000; i++) {
      long tl = Magic.prepareLong(lf, fieldOffset);
      Magic.attemptLong(lf, fieldOffset, tl, l);
      int n0 = (int) tl;
      int n1 = (int) (tl >> 32);
      if (n0 != n1) errors++;
      vi = n;
    }
    tsay(errors + " errors found");
  }

  private static class LongField {
    public long longAccessedViaMagic = 0;
  }

}
