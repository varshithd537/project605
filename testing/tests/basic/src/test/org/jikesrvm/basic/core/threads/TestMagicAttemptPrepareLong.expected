Thread 0 Main: bye
Thread 1 MAPL0: creating
Thread 1 MAPL0: starting
Thread 1 MAPL0: run starting
Thread 1 MAPL0: 0 errors found
Thread 1 MAPL0: run finishing
Thread 2 MAPL1: creating
Thread 2 MAPL1: starting
Thread 2 MAPL1: run starting
Thread 2 MAPL1: 0 errors found
Thread 2 MAPL1: run finishing
Thread 3 MAPL2: creating
Thread 3 MAPL2: starting
Thread 3 MAPL2: run starting
Thread 3 MAPL2: 0 errors found
Thread 3 MAPL2: run finishing
Thread 4 MAPL3: creating
Thread 4 MAPL3: starting
Thread 4 MAPL3: run starting
Thread 4 MAPL3: 0 errors found
Thread 4 MAPL3: run finishing
Thread 5 MAPL4: creating
Thread 5 MAPL4: starting
Thread 5 MAPL4: run starting
Thread 5 MAPL4: 0 errors found
Thread 5 MAPL4: run finishing
